include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,103 +202,138 @@ struct __init_processing
+        if (__iters_per_work_item == 1)
+        {
+            new (&__res.__v) _Tp(__unary_op(__global_idx, __acc...));
+            return;
+        }
+        const _Size __local_range = __item_id.get_local_range(0);
+        const _Size __no_vec_ops = __iters_per_work_item / _VecSize;
         const _Size __adjusted_n = __global_offset + __n;
+        constexpr _Size __vec_size_minus_one = _VecSize - 1;
 
-        // Coalesced load and reduce from global memory
-        if (__adjusted_global_id + __stride * __iters_per_work_item < __adjusted_n)
+        _Size __stride = _VecSize; // sequential loads with _VecSize-wide vectors
+        _Size __adjusted_global_id = __global_offset;
+        if constexpr (_Commutative{})
         {
-            new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            _ONEDPL_PRAGMA_UNROLL
-            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            __stride *= __local_range; // coalesced loads with _VecSize-wide vectors
+            _Size __local_idx = __item_id.get_local_id(0);
+            _Size __group_idx = __item_id.get_group_linear_id();
+            __adjusted_global_id += __group_idx * __local_range * __iters_per_work_item + __local_idx * _VecSize;
+        }
+        else
+            __adjusted_global_id += __iters_per_work_item * __global_idx;
+
+        // If n is not evenly divisible by the number of elements processed per work-group, the last work-group might
+        // need to process less elements than __iters_per_work_item.
+        bool __is_full_wg = __is_full;
+        if (!__is_full_wg)
+        {
+            const bool __is_multi_group = __n_groups > 1;
+            if (__is_multi_group)
+            {
+                _Size __group_idx = __item_id.get_group(0);
+                _Size __n_groups_minus_one = __n_groups - 1;
+                const bool __is_last_wg = __group_idx == __n_groups_minus_one;
+                if (!__is_last_wg)
+                    __is_full_wg = true;
+            }
+        }
+        // _VecSize-wide vectorized path (__iters_per_work_item are multiples of _VecSize)
+        if (__is_full_wg)
+        {
+            vectorized_reduction_first(__adjusted_global_id, __res, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+                vectorized_reduction_remainder(__adjusted_global_id + __i * __stride, __res, __acc...);
         }
+        // At least one vector operation
+        else if (__adjusted_global_id + __vec_size_minus_one < __adjusted_n)
+        {
+            vectorized_reduction_first(__adjusted_global_id, __res, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+            {
+                const _Size __base_idx = __adjusted_global_id + __i * __stride;
+                if (__base_idx + __vec_size_minus_one < __adjusted_n)
```

Comment:
Can we calculate the number of loop iterations outside of the loop? I think that the `scalar_reduction_remainder` call can be pulled after the loop to remove most the branching if we reduce the iteration count by 1 since it should be called on the last iteration if I understand correctly.